### PR TITLE
ramips: mt7621: Add 3 more NETGEAR R6700 v2 clones

### PR DESCRIPTION
**Pull request by dhui**
This PR has 2 changes:

### Add other known clones of the NETGEAR R6700v2.
See original commit adding support:
6e80df5 ("ramips: add support for NETGEAR R6700v2/AC2400")

Add NETGEAR Nighthawk AC2100, NETGEAR R6900 v2, and NETGEAR R7450
as respective DEVICE_ALT2, DEVICE_ALT3, and DEVICE_ALT4
for NETGEAR R6700 v2

See also:
* https://deviwiki.com/wiki/Netgear_R6700_v2
* https://wikidevi.wi-cat.ru/Netgear_R6700_v2

FWIW, I've been running a 21.02 snapshot and 21.02.0-rc2 with this device target (R6700v2) on an AC2100 for over a month now.

### CODE REVIEW

**Comment by adschm** [comment]
Please create the new Rxxxx-named clones as separate devices instead of using DEVICE_ALT.

**Comment by dhui** [comment]
I'm happy to go with either separate devices or `DEVICE_ALT`, but saw that other clones [were](https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=08f5cac6fbd8ac97d6b012a0f6a5f0fabe19c623) [added](https://github.com/openwrt/openwrt/pull/2924) using `DEVICE_ALT`.

Should the existing `DEVICE_ALT`s for the R6700v2 also be made into separate devices?

CC: @CHKDSK88

**Comment by CHKDSK88** [comment]
> 
@adschm  
> Please create the new Rxxxx-named clones as separate devices instead of using DEVICE_ALT.

Could You tell why?

These devices use the same stock firmware file:

```
binwalk /tmp/*img

Scan Time:     2021-08-19 09:30:05
Target File:   /tmp/AC2100-V1.2.0.76_1.0.1.img
MD5 Checksum:  11a6f42627df160f2f29b1f088ae7b92
Signatures:    391

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             Sercomm firmware signature, version control: 256, download control: 0, hardware ID: "BZV", hardware version: 0x4100, firmware version: 0x76, starting code segment: 0x0, code size: 0x7300
512           0x200           Zip archive data, at least v2.0 to extract, compressed size: 33389924, uncompressed size: 83886080, name: R6950.bin
33390564      0x1FD7FE4       End of Zip archive, footer length: 22


Scan Time:     2021-08-19 09:30:05
Target File:   /tmp/R6700v2-V1.2.0.76_1.0.1.img
MD5 Checksum:  11a6f42627df160f2f29b1f088ae7b92
Signatures:    391

DECIMAL       HEXADECIMAL     DESCRIPTION
--------------------------------------------------------------------------------
0             0x0             Sercomm firmware signature, version control: 256, download control: 0, hardware ID: "BZV", hardware version: 0x4100, firmware version: 0x76, starting code segment: 0x0, code size: 0x7300
512           0x200           Zip archive data, at least v2.0 to extract, compressed size: 33389924, uncompressed size: 83886080, name: R6950.bin
33390564      0x1FD7FE4       End of Zip archive, footer length: 22

```

**Comment by adschm** [comment]
> > Please create the new Rxxxx-named clones as separate devices instead of using DEVICE_ALT.
> 
> Could You tell why?
> 

Because the DEVICE_ALT mechanism only works for `make menuconfig` and the metadata-based firmware selector. When you just have a bunch of images, there is no way to tell that r6700-v2 image will also serve r6900-v2 etc.

DEVICE_ALT has been designed for a very special case where devices have both an "internal" and "external" (i.e. marketing) name. Since we have a lot of different (actual) images for the R6xxx series, I think the user experience will be better if we provide images for each individual R6xxx device so the user immediately knows which to choose. The "Nighthawk whatever" may be kept  as DEVICE_ALT; this is roughly the kind of alternate marketing-based name that this feature is actually meant for.

**Comment by dhui** [comment]
> DEVICE_ALT has been designed for a very special case where devices have both an "internal" and "external" (i.e. marketing) name. Since we have a lot of different (actual) images for the R6xxx series, I think the user experience will be better if we provide images for each individual R6xxx device so the user immediately knows which to choose. The "Nighthawk whatever" may be kept as DEVICE_ALT; this is roughly the kind of alternate marketing-based name that this feature is actually meant for.

This makes sense to me. I'll go ahead and create separate devices for these clones.

* Do you want each new device in it's own PR or can I have them all in 1 PR?
* Can you confirm that we'll leave `Nighthawk AC2400` and `Nighthawk AC2100` as `DEVICE_ALT` and move `R7200v1`, `R6900v2`, and `R7450v1` to separate devices?
* For the separate devices, in `target/linux/ramips/image/mt7621.mk`, can I just do `$(Device/netgear_r6700-v2)` and override the model and variant and undefine the DEVICE_ALT? Or do you prefer that I copy, paste and modify `Device/netgear_r6700-v2` for the separate devices?

**Comment by adschm** [comment]
> This makes sense to me. I'll go ahead and create separate devices for these clones.
> 
> * Do you want each new device in it's own PR or can I have them all in 1 PR?

1 PR is fine, but please have a commit per device, so each has proper title and commit description ...
I'd suggest to just force-push to this PR, so we have the discussion around ...

> * Can you confirm that we'll leave `Nighthawk AC2400` and `Nighthawk AC2100` as `DEVICE_ALT` and move `R7200v1`, `R6900v2`, and `R7450v1` to separate devices?

Yes, if those are all clones.

> * For the separate devices, in `target/linux/ramips/image/mt7621.mk`, can I just do `$(Device/netgear_r6700-v2)` and override the model and variant and undefine the DEVICE_ALT? Or do you prefer that I copy, paste and modify `Device/netgear_r6700-v2` for the separate devices?

No, please don't override nodes that are used for an actual device. The proper way is to create another shared node, which will then be used for the devices at hand. The hardest job will then be find a good name for that shared node.

Don't forget to duplicate the DTS files and all those entries in board.d subfiles etc.

**Comment by CHKDSK88** [comment]
Please remember about adding `SUPPORTED_DEVICES ` for easy sysupgrade capability.  Minimum is  `SUPPORTED_DEVICES  += netgear,r6700-v2`

But IMO every device should have other clones in supported devices.

**Comment by adschm** [comment]
> Please remember about adding `SUPPORTED_DEVICES ` for easy sysupgrade capability. Minimum is `SUPPORTED_DEVICES += netgear,r6700-v2`
> 
> But IMO every device should have other clones in supported devices.

I'm against this approach. The idea of separate images is based on having a specific image for a specific device. Cross-flashing should only be enabled for backwards-compatibility. If R6900 is printed on the cover, you should not be able to sysupgrade into an R6700 image.

File: target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh [hunk 1]
```diff
@@ -70,7 +70,8 @@ platform_do_upgrade() {
 	netgear,r6350|\
 	netgear,r6700-v2|\
 	netgear,r7200|\
-	netgear,r6800|\
+	netgear,r6900-v2|\
+        netgear,r6800|\
```
Comment (adschm) [inline target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh]: Broken indent, wrong sorting (sorting seems undefined in several places).

Comment (dhui) [reply target/linux/ramips/mt7621/base-files/lib/upgrade/platform.sh]: I didn't realize these were sorted. I've fixed this now.

**Comment by CHKDSK88** [comment]
> > Please remember about adding `SUPPORTED_DEVICES ` for easy sysupgrade capability. Minimum is `SUPPORTED_DEVICES += netgear,r6700-v2`
> > But IMO every device should have other clones in supported devices.
> 
> I'm against this approach. The idea of separate images is based on having a specific image for a specific device. Cross-flashing should only be enabled for backwards-compatibility. If R6900 is printed on the cover, you should not be able to sysupgrade into an R6700 image.

So,  `SUPPORTED_DEVICES += netgear,r6700-v2` should affect R7200 only?

**Comment by dhui** [comment]
@adschm could you take another look?

**Comment by xsorifc28** [comment]
I would appreciate any efforts to merge this. I have both a R7200 and R7450.
Willing to help test if necessary.

**Comment by adschm** [comment]
Just wanted to take the patches and found that you used a noreply mail address. Please put a proper Signed-off-by here in the comments (!) so we can this done with.

File: target/linux/ramips/image/mt7621.mk [hunk 2]
```diff
@@ -1109,7 +1109,7 @@ TARGET_DEVICES += netgear_r6900-v2
 
 define Device/netgear_r7200
   $(Device/netgear_sercomm_nand)
-  DEVICE_MODEL := R7200
+  DEVICE_MODEL := R7450
```
Comment (adschm) [inline target/linux/ramips/image/mt7621.mk:1112]: That's wrong.

**Comment by adschm** [comment]
For the last two patches, I'd also like to have a proper "device support" commit message, i.e. add specifications and installation instructions for these devices.

File: target/linux/ramips/image/mt7621.mk (hunk 2 above)
Comment (xsorifc28) [reply target/linux/ramips/image/mt7621.mk:1112]: https://github.com/openwrt/openwrt/pull/4408/files#diff-5ac2666a9f7c325aa6d42d94872a91ebe7d8af73e2f708dc0f8564b0f3f88cd8R1124 
Is also wrong, should be swapped?

**Comment by dhui** [comment]
@adschm 

I believe I've addressed all of the outstanding issues:
* Use an actual email address
* Fixed swapped `DEVICE_MODEL` in `target/linux/ramips/image/mt7621.mk`
* Provided spec and install instructions for the 2 new devices

**Comment by adschm** [comment]
Last commit still contained the wrong compatible/model in DTS, but since you had to jump above a lot of obstacles already, I'll fix that myself and merge this during the weekend. Thanks for keeping up.

**Comment by dhui** [comment]
Thanks @adschm !

**Comment by xsorifc28** [comment]
I'd like to test openwrt on my R7450. Curious to know when this build will be available. Is there a timeline or any documentation on how to test?

**Comment by dhui** [comment]
@xsorifc28 I've been using the [21.02.1 release for the R6700v2](https://firmware-selector.openwrt.org/?version=21.02.1&target=ramips%2Fmt7621&id=netgear_r6700-v2) on my AC2100 without any issues so it _should_ work for your R7450 as well. But if you're feeling [adventurous](https://openwrt.org/docs/guide-quick-start/developmentinstallation) you can try the [R7450 snapshot](https://firmware-selector.openwrt.org/?version=SNAPSHOT&target=ramips%2Fmt7621&id=netgear_r7450)

**Comment by xsorifc28** [comment]
Thanks a lot, appreciated.

**Pull request closed without merging**